api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/computation/ExecutionTree.java

```diff
@@ -0,0 +1,98 @@
+package edu.cornell.mannlib.vitro.webapp.dynapi.computation;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class ExecutionTree {
+    private Map<StepInfo, List<StepInfo>> forwardMap;
+    private Map<StepInfo, List<StepInfo>> backwardMap;
+    
+    public ExecutionTree() {
+        //map from previous to list of next
+        forwardMap = new HashMap<StepInfo, List<StepInfo>>();
+        //map from next to list of previous
+        backwardMap = new HashMap<StepInfo, List<StepInfo>>();
+    }
+    
+    public ExecutionTree(StepInfo rootNode) {
+        this();
+        initialize(rootNode);
+    }
+
+    private void initialize(StepInfo prev) {
+        if (prev == null) {
+            return;
+        }
+        Set<StepInfo> nextNodes = prev.getNextNodes();
+        for (StepInfo next : nextNodes) {
+            if (!containsPrev(next)) {
+                addArc(prev, next);
+                initialize(next);
+            }
+        }
+    }
+    
+    public List<StepInfo> getLeafs(){
+        return getPrevOf(null);
```

Comment:
This appears to be asking for key null in backwardMap. Which hopefully, will always be null. If not, do we really want to operate with null keys?

Reply:
Fixed by introduction NullStep